### PR TITLE
Adding support for directories/packagePath in package.json.

### PR DESCRIPTION
**Pull request by dzearing**
This change allows projects to specify they build their package output in a specific directory. This allows us to make major optimizations in npmx workflows, like unpacking the packages to prepopulate the build output.

It also includes a prototype build receipt task, which is actually 2 tasks: a check step and an update step. The check step evaluates if files in the project have changed since last built. It does this by looking at git sha1 hashes for files and comparing them hashes in a build.json file published in the package folder. If the hashes match, the remaining subtasks are skipped. If they dont, the rest of the subtasks run and the update step runs last, which updates the build.json with the latest hashes.

That means that if package files haven't changed since last published, we can avoid building completely.

I'd like to also include the change detection tasks, but won't modify the build rigs to include it until we've sorted out the right layer to include it. Per discussion, right now the git hash logic takes 50-100ms for a project. We're looking at moving this logic to a higher level within NPMX, so that we can further optimize completely avoiding spinning up processes to load gulp and all that stuff. Right now just to load up gulp in a quadcore i7 macbook pro, it takes 3 seconds. Multiplied times 45 packages, we end up spending 135 seconds just loading gulp on a pure warm build. So if we can reduce that down to 50-100ms per package, we're looking at slimming this down to maybe 4 seconds of overhead.

### CODE REVIEW

File: gulp-core-build/src/IExecutable.ts [hunk 1]
```diff
@@ -8,7 +8,7 @@ export interface IExecutable {
   name?: string;
 
   /** Optional callback to indicate if the task is enabled or not. */
```
Comment (cliffkoh) [inline gulp-core-build/src/IExecutable.ts]: For props like this which takes in callbacks, good to document what kind of callback signature is expected..

File: gulp-core-build/src/tasks/BuildReceiptTask.ts [hunk 2]
```diff
@@ -0,0 +1,134 @@
+import { GulpTask } from './GulpTask';
+import gulp = require('gulp');
+import * as fs from 'fs';
+import * as path from 'path';
+import * as child_process from 'child_process';
+
+export interface IBuildReceiptTask {
+}
+
+const PROCESS_OUTPUT_DELIMITER: string = '~X~X~X~X~X~X~';
```
Comment (cliffkoh) [inline gulp-core-build/src/tasks/BuildReceiptTask.ts]: Might be good to use in the delimiter here, symbols which are disallowed in file paths

Comment (cliffkoh) [inline gulp-core-build/src/tasks/BuildReceiptTask.ts]: For example, consecutive slashes ( ~X///~X%X~\\\X~)

File: gulp-core-build/src/tasks/BuildReceiptTask.ts [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+let _lastLocalHashes: { [path: string]: string } = {};
+
+/**
+ * This task is responsible for the following:
+ *
+ * 1. Gather the local files build receipt of the current files. Cache it in _lastFilesHash for later use.
```
Comment (cliffkoh) [inline gulp-core-build/src/tasks/BuildReceiptTask.ts]: "Gather the local files build receipt of...", "local files build receipt" is a little hard to parse...

File: gulp-core-build/src/tasks/BuildReceiptTask.ts [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+ *
+ * 2. Check for the existence of a package build receipt. If none exists, complete the task.
+ *
+ * 3. Compare the package build hashes with the local build hashes. If they match, update the buildConfig to
+ *    have isRedundantBuild flat set to true. This allows other tasks to completely prematurely.
```
Comment (cliffkoh) [inline gulp-core-build/src/tasks/BuildReceiptTask.ts]: flag

File: test-web-library-build/.vscode/launch.json [hunk 5]
```diff
@@ -0,0 +1,46 @@
+{
```
Comment (cliffkoh) [inline test-web-library-build/.vscode/launch.json]: Err, what are these? Do you actually intend to check these in?

File: gulp-core-build/src/tasks/BuildReceiptTask.ts [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+export class CheckBuildReceiptTask extends GulpTask<IBuildReceiptTask> {
+  public name: string = 'check-for-changes';
+  public executeTask(
+    gulp: gulp.Gulp,
+    completeCallback: (result?: Object) => void
+  ): Promise<Object> | NodeJS.ReadWriteStream | void {
+
+    getLocalHashes().then(localHashes => {
+      _lastLocalHashes = localHashes;
+      debugger;
```
Comment (cliffkoh) [inline gulp-core-build/src/tasks/BuildReceiptTask.ts]: debugger statement here

File: gulp-core-build/src/tasks/BuildReceiptTask.ts [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+      readPackageHashes(path.join(process.cwd(), this.buildConfig.packageFolder, 'build.json')).then(packageHashes => {
+        if (packageHashes) {
+          if (areObjectsEqual(localHashes, packageHashes)) {
+            this.buildConfig.isRedundantBuild = true;
+            this.log('Build is redundant. Skipping steps.');
+          } else {
+            debugger;
```
Comment (nickpape-msft) [inline gulp-core-build/src/tasks/BuildReceiptTask.ts]: remove
